File: src/11/11.rs
use std::collections::HashMap;
use std::io::{self, Read};

fn split_number(n: u64) -> (u64, u64) {
    let s = n.to_string();
    let mid = s.len() / 2;
    let left = s[..mid].parse::<u64>().unwrap();
    let right = s[mid..].parse::<u64>().unwrap();
    (left, right)
}

fn transform_stones(stones: Vec<u64>, steps: u32) -> u64 {
    let mut stones_dict: HashMap<u64, u64> = HashMap::new();
    for &stone in &stones {
        *stones_dict.entry(stone).or_insert(0) += 1;
    }

    for _ in 0..steps {
        let mut new_stones: HashMap<u64, u64> = HashMap::new();
        for (&stone, &count) in &stones_dict {
            if stone == 0 {
                *new_stones.entry(1).or_insert(0) += count;
            } else {
                let s = stone.to_string(); // Correctly get the string representation of the stone
                if s.len() % 2 == 0 {
                    let (left, right) = split_number(stone);
                    *new_stones.entry(left).or_insert(0) += count;
                    *new_stones.entry(right).or_insert(0) += count;
                } else {
                    *new_stones.entry(stone * 2024).or_insert(0) += count;
                }
            }
        }
        stones_dict = new_stones;
    }

    stones_dict.values().sum()
}

fn main() -> io::Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let stones: Vec<u64> = input.trim().split_whitespace().map(|s| s.parse::<u64>().unwrap()).collect();

    println!("{}", transform_stones(stones.clone(), 25));
    println!("{}", transform_stones(stones, 75));

    Ok(())
}

File: 2024/11/11.py
#!/usr/bin/env python3
import sys
from collections import defaultdict
from typing import List, Dict

def split_number(n: int) -> (int, int):
    """Split a number into two halves."""
    s = str(n)
    mid = len(s) // 2
    left = int(s[:mid])
    right = int(s[mid:])
    return left, right

def transform_stones(stones: List[int], steps: int = 25) -> int:
    """Transform the list of stones according to the rules using lazy evaluation and counting."""
    # Initialize the stones dictionary
    stones_dict: Dict[int, int] = defaultdict(int)
    for stone in stones:
        stones_dict[stone] += 1
    
    for _ in range(steps):
        new_stones: Dict[int, int] = defaultdict(int)
        for stone, count in stones_dict.items():
            if stone == 0:
                new_stones[1] += count
            elif len(str(stone)) % 2 == 0:
                left, right = split_number(stone)
                new_stones[left] += count
                new_stones[right] += count
            else:
                new_stones[stone * 2024] += count
        stones_dict = new_stones
    
    return sum(count for stone, count in stones_dict.items())

def main():
    # Read input from stdin
    stones: List[int] = list(map(int, sys.stdin.read().strip().split()))
    print(transform_stones(stones, steps=25))
    print(transform_stones(stones, steps=75))

if __name__ == "__main__":
    main()